### PR TITLE
*: supports Unifiled Log Format

### PR DESCRIPTION
**Pull request by nolouch**
Signed-off-by: nolouch <nolouch@gmail.com>
supports [Unified Log Format](https://github.com/tikv/rfcs/blob/master/text/2018-12-19-unified-log-format.md)

### CODE REVIEW

File: log.go [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+	core := zapcore.NewCore(newZapTextEncoder(cfg), output, level)
+	lg := zap.New(core, cfg.buildOptions(output)...)
+
+	return lg, output, nil
+}
+
+// initFileLog initializes file based logging options.
+func initFileLog(cfg *FileLogConfig) (*lumberjack.Logger, error) {
+	if st, err := os.Stat(cfg.Filename); err == nil {
+		if st.IsDir() {
```
[the remark this answers is not in the document]
Comment (nolouch) [reply log.go]: I think it will processed by `lumberjack`.

File: zap_text_encoder.go [hunk 2]
```diff
@@ -0,0 +1,608 @@
...
+	return &textEncoder{}
+}}
+
+var (
+	_pool = buffer.NewPool()
+	// Get retrieves a buffer from the pool, creating one if necessary.
+	Get = _pool.Get
+)
+
+func getTextEncoder() *textEncoder {
```
Comment (breeswish) [inline zap_text_encoder.go]: why do we need a pool for the encoder?

File: zap_text_encoder.go [hunk 3]
```diff
@@ -0,0 +1,608 @@
...
+	ne.buf.AppendByte(']')
+	enc.AppendByteString(ne.buf.Bytes())
+	ne.buf.Free()
+	putTextEncoder(ne)
+	return err
+}
+
+func (enc *textEncoder) AppendObject(obj zapcore.ObjectMarshaler) error {
+	enc.addElementSeparator()
+	ne := enc.cloned()
```
Comment (breeswish) [inline zap_text_encoder.go]: can this be pooled?

File: zap_text_encoder.go (hunk 2 above)
Comment (nolouch) [reply zap_text_encoder.go]: Sometimes we need `cloned` to create a temp encoder, so use a pool to reduce allocation.

File: zap_text_encoder.go (hunk 3 above)
Comment (nolouch) [reply zap_text_encoder.go]: Yes.

**Pull request merged**